src/fparser/one/tests/test_do.py

```diff
@@ -0,0 +1,212 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2019 Science and Technology Facilities Council.
+#
+# All rights reserved.
+#
+# Modifications made as part of the fparser project are distributed
+# under the following license:
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+# 1. Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+'''
+Tests all permutations of the DO block. There are a lot of them.
+'''
+from __future__ import absolute_import, print_function
+
+import pytest
+
+from fparser.common.utils import AnalyzeError
+from fparser.common.sourceinfo import FortranFormat
+from fparser.one.parsefortran import FortranParser
+from fparser.common.readfortran import FortranStringReader
+
+
+@pytest.mark.parametrize('name', [None, 'loop_name'])
+@pytest.mark.parametrize('label', [None, '123'])
+@pytest.mark.parametrize('control_comma', [False, True])
+@pytest.mark.parametrize('initial_expression',
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('terminal_expression',
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('incrament_expression',
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('end_name', [None, 'loop_name', 'wrong_name'])
+@pytest.mark.parametrize('end_label', [None, '123', '456'])
+def test_do(name, label, control_comma, initial_expression,
```

Comment:
Please name these tests for the rules they are checking and/or document this in the docstrings.